libs/discord/discord.cpp

```diff
@@ -0,0 +1,212 @@
+{
+	Com_GGPrint("Discord: [{}]: {}", log2str(level), message);
+}
+
+static Application app = {};
+static IDiscordUserEvents users_events = {};
+static IDiscordActivityEvents activities_events = {};
+
+
+void InitDiscord()
+{
+	ZoneScoped;
+
+	discord_sdk = LoadLibraryW(L"discord_game_sdk.dll");
+	if (discord_sdk == NULL)
+		return;
+
+	pDiscordCreate = (tDiscordCreate)GetProcAddress(discord_sdk, "DiscordCreate");
+	if (pDiscordCreate == NULL)
+		return;
+
+
+	users_events.on_current_user_update = OnUserUpdated;
+
+	DiscordCreateParams params;
+	DiscordCreateParamsSetDefault(&params);
+	params.client_id = CLIENT_ID;
+	params.flags = DiscordCreateFlags_NoRequireDiscord;	// Fail if discord is not running instead of closing the game
+	params.event_data = &app;
+	params.user_events = &users_events;
+	params.activity_events = &activities_events;
+	EDiscordResult result = pDiscordCreate(DISCORD_VERSION, &params, &app.core);
+	if (result != DiscordResult_Ok) {
+		Com_GGPrint("Discord: not initialized({})", result);
+		// Be sure it's disabled
+		app.core = nullptr;
+	} else {
+		app.core->set_log_hook(app.core, DiscordLogLevel_Debug, nullptr, OnLog);
+
+		app.users = app.core->get_user_manager(app.core);
+		app.activities = app.core->get_activity_manager(app.core);
+	}
+}
+
+
+static connstate_t old_state = CA_UNINITIALIZED;
+static int old_team = -1;
+static int old_match_state = -1;
+static int old_round = -1;
+
+
+void DiscordFrame()
+{
+	ZoneScoped;
+
+	if (app.activities)
+	{
+		if( (cls.state == CA_CONNECTED || cls.state == CA_ACTIVE) && client_gs.module )
+		{
+
+			if (old_state != cls.state ||
+				old_team != cg.predictedPlayerState.real_team ||
+				old_match_state != client_gs.gameState.match_state ||
+				old_round != client_gs.gameState.round_num)
+			{
+				old_state = cls.state;
+				old_team = cg.predictedPlayerState.real_team;
+				old_match_state = client_gs.gameState.match_state;
+				old_round = client_gs.gameState.round_num;
+
+				DiscordActivity activity = {};
+				//activity.type = DiscordActivityType_Playing;
+
+				if (cg.predictedPlayerState.real_team == TEAM_SPECTATOR) {
+					Q_strncpyz(activity.details, "SPECTATING");
+				} else if (client_gs.gameState.match_state <= MatchState_Warmup) {
+					Q_strncpyz(activity.details, "WARMUP");
+				} else if (client_gs.gameState.match_state <= MatchState_Playing) {
+					ggformat(activity.details, "ROUND {}", client_gs.gameState.round_num);
+				}
+
+				bool is_bomb = GS_TeamBasedGametype( &client_gs );
+				if (is_bomb) {
+					auto& alpha = client_gs.gameState.teams[ TEAM_ALPHA ];
+					auto& beta = client_gs.gameState.teams[ TEAM_BETA ];
+					ggformat(activity.state, "bomb: {}-{}", alpha.score, beta.score );
+				} else {
+					Q_strncpyz(activity.state, "gladiator" );
+				}
+
+				if (client_gs.gameState.match_state_start_time) {
+					time_t now = time( NULL );
+					activity.timestamps.start = now - client_gs.gameState.match_state_start_time / 1000;
```

Comment:
I think match_state_start_time counts from 0 when the server changes map

not 100% sure what the right timestamp to use here is but it's not `time`